src/main/scala/mesosphere/marathon/state/Container.scala

```diff
@@ -3,75 +3,53 @@ package mesosphere.marathon.state
 import com.wix.accord.dsl._
 import com.wix.accord._
 import mesosphere.marathon.api.v2.Validation._
-import org.apache.mesos.{ Protos => Mesos }
+
+import org.apache.mesos.Protos.ContainerInfo
+
 import scala.collection.immutable.Seq
 
-// TODO: trait Container and specializations?
-// Current implementation with type defaulting to DOCKER and docker to NONE makes no sense
-case class Container(
-    `type`: Mesos.ContainerInfo.Type = Mesos.ContainerInfo.Type.DOCKER,
-    volumes: Seq[Volume] = Nil,
-    docker: Option[Container.Docker] = None) {
+trait Container {
+  val volumes: Seq[Volume]
+
+  def foldDocker[R](defaultResult: R)(f: Container.Docker => R): R = {
+    this match {
+      case docker: Container.Docker => f(docker)
+      case _ => defaultResult
+    }
+  }
 
-  import Container._
+  def docker(): Option[Container.Docker] = foldDocker[Option[Container.Docker]](None)(Some(_))
```

Comment:
Do we really need this? Will the code where this is used not be more readable if we match against the type?